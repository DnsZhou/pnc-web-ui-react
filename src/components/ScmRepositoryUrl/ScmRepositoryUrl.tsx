import { Button } from '@patternfly/react-core';
import { ExternalLinkAltIcon } from '@patternfly/react-icons';

import { CopyToClipboard } from 'components/CopyToClipboard/CopyToClipboard';
import { TooltipWrapper } from 'components/TooltipWrapper/TooltipWrapper';

import { parseInternalRepositoryUrl } from 'utils/utils';

interface IGerritButtonProps {
  url: string;
  isInline?: boolean;
}

/**
 * The Gerrit button that redirect to specific Gerrit page.
 *
 * @param url - the internal/external url for the SCM Repository
 * @param isInline - whether to use inline style with external link action
 */
const GerritButton = ({ isInline, url }: IGerritButtonProps) => (
  <TooltipWrapper tooltip="View in Gerrit">
    <Button
      component="a"
      href={parseInternalRepositoryUrl({ internalUrl: url })}
      target="_blank"
      rel="noopener noreferrer"
      variant={isInline ? 'plain' : 'tertiary'}
      icon={<ExternalLinkAltIcon />}
    >
      {isInline ? <ExternalLinkAltIcon /> : 'Gerrit'}
    </Button>
  </TooltipWrapper>
);

interface IScmRepositoryUrlProps {
  url: string;
  showClipboardCopy?: boolean;
  showGerritButton?: boolean;
  isInline?: boolean;
}

/**
 * Represents a URL to a ProjectDetailPage of a specific project.
 *
 * @param url - the internal/external url for the SCM Repository
 * @param showClipboardCopy - whether to display the url as clipboard copy
 * @param showGerritButton - whether to display the Gerrit button
 * @param isInline - whether to use inline style with external link action
 */
export const ScmRepositoryUrl = ({ url, showClipboardCopy = true, showGerritButton, isInline }: IScmRepositoryUrlProps) => {
  if (showClipboardCopy) {
    return (
      <CopyToClipboard isInline={isInline} suffixComponent={showGerritButton && <GerritButton isInline={isInline} url={url} />}>
        {url}
      </CopyToClipboard>
    );
  }
  return (
    <span>
      {url}
      {showGerritButton && <GerritButton isInline={isInline} url={url} />}
    </span>
  );
};
