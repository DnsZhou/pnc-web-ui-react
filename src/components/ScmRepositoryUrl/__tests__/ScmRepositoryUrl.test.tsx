import { render } from '@testing-library/react';

import { ScmRepositoryUrl } from 'components/ScmRepositoryUrl/ScmRepositoryUrl';

describe('display ScmRepositoryUrl component', () => {
  test('renders ScmRepositoryLink', () => {
    render(
      <ScmRepositoryUrl
        isInternal
        scmRepository={{
          id: '103',
          internalUrl: 'git+ssh://code.test.env.com/testRepo/testUrlClipboardCopyGerrit.git',
          externalUrl: 'https://github.com/testRepo/empty.git',
          preBuildSyncEnabled: true,
        }}
      />
    );
  });

  test('compare snapshot with previous record', () => {
    const tree = render(
      <ScmRepositoryUrl
        isInternal
        scmRepository={{
          id: '103',
          internalUrl: 'git+ssh://code.test.env.com/testRepo/testUrlClipboardCopyGerrit.git',
          externalUrl: 'https://github.com/testRepo/empty.git',
          preBuildSyncEnabled: true,
        }}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
