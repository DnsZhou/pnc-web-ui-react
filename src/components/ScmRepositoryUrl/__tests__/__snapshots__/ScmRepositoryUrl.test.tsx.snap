// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display ScmRepositoryUrl component compare snapshot with previous record 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span>
        git+ssh://code.test.env.com/testRepo/testUrlClipboardCopyGerrit.git
        <div
          style="display: contents;"
        >
           
          <a
            aria-disabled="false"
            class="pf-c-button pf-m-tertiary"
            data-ouia-component-id="OUIA-Generated-Button-tertiary-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            href="https://code.test.env.com/gerrit/gitweb?p=testRepo/testUrlClipboardCopyGerrit.git;a=summary"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="pf-c-button__icon pf-m-start"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                />
              </svg>
            </span>
            Gerrit
          </a>
        </div>
      </span>
    </div>
  </body>,
  "container": <div>
    <span>
      git+ssh://code.test.env.com/testRepo/testUrlClipboardCopyGerrit.git
      <div
        style="display: contents;"
      >
         
        <a
          aria-disabled="false"
          class="pf-c-button pf-m-tertiary"
          data-ouia-component-id="OUIA-Generated-Button-tertiary-2"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe="true"
          href="https://code.test.env.com/gerrit/gitweb?p=testRepo/testUrlClipboardCopyGerrit.git;a=summary"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="pf-c-button__icon pf-m-start"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
              />
            </svg>
          </span>
          Gerrit
        </a>
      </div>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
